### ISSUE TITLE
Cannot change keyboard layout/language

### ISSUE THREAD

**Issue by CrisesUrv** [libretro/RetroArch#13838]
Retroarch uses a US English keyboard layout as default (tested by setting keys in the GUI or in cores with direct keyboard input), independently of what layout is set in the OS (at least in linux/Raspbian). 

It does not seem there is an option in the GUI to set the keyboard layout, but there is an option in the retroarch.cfg file that says:

`# Keyboard layout for input driver if applicable (udev/evdev for now).
# Syntax is either just layout (e.g. "no"), or a layout and variant separated with colon ("no:nodeadkeys").
# input_keyboard_layout =`

However, no matter what keyboard code is set (for example 'es' for Spanish), it does not seem to have an effect: the US keyboard layout is always used.

Tested on:
Retroarch 1.10
Retropie 4.8 on a Raspberry pi 4

**Comment by keithbowes** [comment]
Yeah, it's never worked for me either (I still have it set to `input_keyboard_layout = "us:dvorak"`, with no effect).    I tried to change the udev driver once but couldn't wrap my head around exactly where the XKB functions were supposed to do their magic.

I just gave up and started using physical rather than logical keys in my config/remaps.  I have trying again on my TODO list, but it's way down on that list.

**Comment by sonninnos** [comment]
Why would you need to do it at the frontend level though, when it is possible to change the layout in all (?) computer cores that actually use a keyboard..?

**Comment by CrisesUrv** [comment]
> Why would you need to do it at the frontend level though, when it is possible to change the layout in all (?) computer cores that actually use a keyboard..?

Well, there are some corner cases. For example, Flycast supports the dreamcast keyboard but does not allow changing the layout. On the other hand, retroarch hotkeys mapped to the keyboard will also follow the US layout rather than the actual keyboard layout.

**Comment by sonninnos** [comment]
All Windows drivers currently follow the physical layout, so that OS locale does not matter. "Raw" driver used to act so that user would have to switch to US layout in order to get correct keys in cores. I use FI layout, so having to switch it constantly is/was annoying.

Surely there ought to be at least the option to get the same behavior in other platform drivers too.

**Comment by MeisterLLD** [comment]
Any news on that ? Did OP find a solution ?
I just posted this feature request on flycast

**Comment by MeisterLLD** [comment]
I just checked, same behaviour happens on a Raspberry Pi 4 with Retropie, but also on my Arch Linux with Retroarch 1.15 : bind some control to "A" on a properly configured azerty keyboard, with or without 
`input_keyboard_layout = "fr" `
the behaviour is the same : binding is shown as "q" and not "a".

It would be really nice to activate this feature. Since the option exists in the config file it seems that it was planned at some time ?

If one just uses the keyboard for bindings, this problem can be completely bypassed, but when one uses the keyboard to emulate... physical keyboards (e.g. Dreamcast), then this is real trouble as it is impossible to get another layout than us !

**Comment by dante740** [comment]
I've found the answer.

1. download and install 'autohotkey'.

2. create a new script.

3. Paste the following:




#InstallKeybdHook
;Converts Dvorak keystrokes to their correct letters on a QWERTY-configured OS or vice versa

;Script instructions:

;ScrollLock toggles QWERTY or Dvorak mode

;Shift-Caps works like regular Caps

SetKeyDelay, 1, 50

ScrollLock::Suspend, Toggle
-::[
=::]
q::'
w::,
e::.
r::p
t::y
y::f
u::g
i::c
o::r
p::l
[::/
]::=
;\::\ no change
;a::a no change
s::o
d::e
f::u
g::i
h::d
j::h
k::t
l::n
SC027::s
SC028::-
z::SC027
x::q
c::j
v::k
b::x
n::b
m::m
,::w
.::v
/::z




4. Save the script.

5. SET YOUR KEYBOARD TO QWERTY MODE.

6. run the script.

. (you might need to run the script as administrator)

Enjoy.
